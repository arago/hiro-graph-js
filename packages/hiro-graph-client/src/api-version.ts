const APP_API_BASE = '/api/app/7.0';
const AUTH_API_BASE = '/api/auth/6.1';
const EVENTS_WS_API_BASE = '/api/events-ws/6.1';
const GRAPH_API_BASE = '/api/graph/7.1';
const GRAPH_WS_API_BASE = '/api/graph-ws/6.1';
const IAM_API_BASE = '/api/iam/6.1';
const KI_API_BASE = '/api/ki/6';
const VARIABLES_API_BASE = '/api/variables/6';

export {
  APP_API_BASE,
  AUTH_API_BASE,
  EVENTS_WS_API_BASE,
  GRAPH_API_BASE,
  GRAPH_WS_API_BASE,
  IAM_API_BASE,
  KI_API_BASE,
  VARIABLES_API_BASE,
};
