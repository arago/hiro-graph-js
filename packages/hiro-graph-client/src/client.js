/**
 *  A GraphIT Client
 *
 *  It will use Websockets where possible and fall back to HTTP where WebSockets not supported.
 *  Event streams require WebSockets so those will not work witout websocket support.
 *  The connection requires a `Token`.
 */
import WebsocketTransport, { webSocketsAvailable } from './transport-websocket';
import HttpTransport from './transport-http';
import {
    isUnauthorized,
    isTransactionFail,
    isConflict,
    isNotFound,
} from './errors';
import { fixedToken } from './token';
import EventStream from './eventstream';
import authServlet from './servlets/auth';
import apiServlet from './servlets/api';

const passthru = (fn) => [
    (r) => (fn(), r),
    (e) => {
        fn();
        throw e;
    },
];

//nb this isn't a deep clone, just a top level deref.
//we only deal with primitives, arrays and plain objects.
const dereference = (obj) => {
    if (typeof obj !== 'object' || obj === null) {
        return obj;
    }

    if (Array.isArray(obj)) {
        return obj.slice();
    }

    return Object.assign({}, obj);
};

export default class Client {
    constructor({ endpoint, token }, transportOptions = {}, proxies = []) {
        this.endpoint = endpoint;

        //we hold on to the token for ease of access/manual invalidation
        this.setToken(token);

        //create the transports.
        //the http is for http only endpoint, e.g. the /_ki/* and /_variable/* servlets
        this.http = new HttpTransport(endpoint);

        // This is the main transport for most of the api.
        // Use Websocket if available (except if specifically disabled)
        // fallback to http, or use given transport
        // duck type the final parameter
        if (typeof transportOptions.request === 'function') {
            this.transport = transportOptions;
        } else if (webSocketsAvailable && !transportOptions.forceHTTP) {
            //All a transport needs to implement is "request"
            this.transport = new WebsocketTransport(endpoint, transportOptions);
        } else {
            if (!webSocketsAvailable) {
                console.warn(
                    'WebSockets not available, falling back to HTTP transport',
                );
            }

            this.transport = this.http;
        }

        // Bind our fetch for extension servlets.
        this.fetch = (...args) => this.http.fetch(this.token, ...args);
        this.proxyFetch = (proxy) => (url, ...args) =>
            this.fetch(`${proxy}${url}`, ...args);

        this._dedup = Object.create(null);

        //keep this so we can duplicate them
        this._servlets = []; // todo: change _servlets to PlainObject. move custom servlets from this to this._servlets. check getters and setters

        // Auth API
        this.addServlet(
            'auth',
            authServlet,
            proxies.length >= 1 ? proxies[0] : '',
        );

        // Global API
        this.addServlet(
            'api',
            apiServlet,
            proxies.length >= 2 ? proxies[1] : '',
        );
    }

    setToken(token) {
        this.token = typeof token === 'string' ? fixedToken(token) : token;
    }

    // NB this is not held anywhere in this instance, but returned
    // to the caller. It only connects when it's subscribe() method
    // is called.
    eventStream(filters = [], { groupId, offset, scopeId } = {}) {
        let filtersArray = filters;

        if (!Array.isArray(filters)) {
            filtersArray = [filters];
        }

        if (filtersArray.some((f) => typeof f !== 'string')) {
            throw new Error(
                'All filters must be strings representing `jfilter` filters',
            );
        }

        if (filtersArray.length === 0) {
            // add a default one that catches everything
            filtersArray.push('(element.ogit/_id = *)');
        }

        return new EventStream(
            { endpoint: this.endpoint, token: this.token },
            { groupId, offset, scopeId, filters: filtersArray },
        );
    }

    /**
     *  Make a clone of this connection, but with new
     */
    cloneWithNewToken(newToken) {
        const cloned = new Client(
            {
                endpoint: this.endpoint,
                token: newToken,
            },
            this.transport,
        );

        this._servlets.forEach(([name, methods]) =>
            cloned.addServlet(name, methods),
        );

        return cloned;
    }

    /**
     *  Make a generic request. The transport will handle the details, we handle the retry.
     *  We use `this.transport` which is websocket if available.
     */
    request(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1,
    ) {
        return this.transport.request(
            this.token,
            { type, headers, body },
            reqOptions,
        );
    }

    // Deduplicates a request, but the request must be an Array returning one.
    dedupedRequest(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1,
    ) {
        try {
            const requestKey = JSON.stringify({ type, headers, body });

            if (requestKey in this._dedup) {
                this._dedup[requestKey]._calls++;

                return this._dedup[requestKey].then(dereference);
            }

            const cleanUp = passthru(() => {
                const calls = this._dedup[requestKey]._calls;

                delete this._dedup[requestKey];
            });
            const promise = (this._dedup[requestKey] = this.request(
                { type, headers, body },
                reqOptions,
                retries,
            )
                .then(...cleanUp)
                .then((res) => (promise._calls > 1 ? dereference(res) : res)));

            promise._calls = 1;

            return promise;
        } catch (e) {
            return Promise.reject(e);
        }
    }

    getToken() {
        return this.token;
    }

    /**
     *  Core GraphIT REST API methods
     *  Prefer websocket transport
     *  The servlets for e.g. /_ki/validate are added using mixins
     */

    /**
     *  Get a single item by ID
     */
    get(id, reqOptions = {}) {
        const headers = { 'ogit/_id': id };

        if (reqOptions.listMeta) {
            headers.listMeta = true;
        }

        return this.dedupedRequest(
            { type: 'get', headers },
            reqOptions,
        );
    }

    /**
     *  Get the node for the owner of this token
     */
    me(reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'getme',
                body: {},
                headers: { profile: true, 'me-type': 'account' },
            },
            reqOptions,
        );
    }

    /**
     *  Create a new node
     */
    create(type, data = {}, reqOptions = {}) {
        const headers = { 'ogit/_type': type };

        if (reqOptions.waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request(
            { type: 'create', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Update a Node
     */
    update(id, data = {}, reqOptions = {}) {
        const headers = { 'ogit/_id': id };

        if (reqOptions.waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request(
            { type: 'update', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Replace a Node, optionally `upsert`
     */
    replace(id, data = {}, reqOptions = {}) {
        const { createIfNotExists = false, waitForIndex = false } = reqOptions;
        //createIfNotExists should contain the "ogit/_type" to create if the node doesn't exist,
        //and nothing otherwise
        const headers = { 'ogit/_id': id };

        if (createIfNotExists) {
            Object.assign(headers, {
                createIfNotExists: 'true',
                'ogit/_type': createIfNotExists,
            });
        }

        if (waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return his.request(
            { type: 'replace', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Delete a node/edge
     *
     *  We don't handle the 404/409 here. or should we?
     */
    delete(id, reqOptions = {}) {
        const { waitForIndex = false } = reqOptions;
        const headers = { 'ogit/_id': id };

        if (waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request({ type: 'delete', headers }, reqOptions);
    }

    /**
     *  This is a vertices query
     */
    lucene(
        query = '',
        {
            limit = 50,
            offset = 0,
            order = false,
            fields = [],
            count = false,
            ...placeholders
        } = {},
        reqOptions = {},
    ) {
        const body = {
            query,
            limit,
            offset,
            ...placeholders,
        };

        if (count) {
            body.count = 'true'; // string true
        }

        if (order) {
            body.order = '' + order; // this implicitly does array.join(",") for arrays. (and works with strings...)
        }

        if (fields.length > 0) {
            body.fields = Array.isArray(fields)
                ? fields.join(',')
                : String(fields);
        }

        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'vertices' },
                body,
            },
            reqOptions,
        );
    }

    ids(list, reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'ids' },
                body: { query: list.join(',') }, // yes, it has to be a comma-seperated string
            },
            reqOptions,
        );
    }

    /**
     *  This is a gremlin query
     */
    gremlin(root, query, reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'gremlin' },
                body: { root, query },
            },
            reqOptions,
        );
    }

    /**
     *  Connect two Nodes with an edge of `type`
     */
    connect(type, inId, outId, reqOptions = {}) {
        returnthis
            .request(
                {
                    type: 'connect',
                    headers: { 'ogit/_type': type },
                    body: { in: inId, out: outId },
                },
                reqOptions,
            )
            .then(
                () => {}, //return nothing.
                (err) => {
                    //Conflict is OK here, just means that the edge was already connected.
                    if (isConflict(err)) {
                        return; //return nothing.
                    }

                    //real error.
                    throw err;
                },
            );
    }

    /**
     *  Disconnect two nodes, convenience for delete, generates the edge id for you.
     */
    disconnect(type, inId, outId, reqOptions = {}) {
        return this.delete(`${outId}$$${type}$$${inId}`, reqOptions).then(
            () => {}, //return nothing.
            (err) => {
                //Not Found or Conflict is OK here, just means that the edge was already deleted/didn't ever exist
                if (isNotFound(err) || isConflict(err)) {
                    return; //return nothing.
                }

                //real error.
                throw err;
            },
        );
    }

    /**
     *  Write timeseries values (only ogit/Timeseries vertices)
     *
     *  values are { timestamp: millisecond unix, value: string value }
     */
    writets(timeseriesId, values) {
        let items = values;

        if (!Array.isArray(values)) {
            items = [values];
        }

        return this.request({
            type: 'writets',
            headers: {
                'ogit/_id': timeseriesId,
            },
            body: { items },
        });
    }

    /**
     *  Read timeseries values (only ogit/Timeseries vertices)
     */
    streamts(
        timeseriesId,
        opts = {
            from: false,
            to: false,
            limit: 50,
            includeDeleted: false,
            with: [],
        },
    ) {
        const headers = {
            'ogit/_id': timeseriesId,
            ...opts,
        };
        const body = {};

        return this.dedupedRequest({
            type: 'streamts',
            headers,
            body,
        });
    }

    /**
     * Placeholders here for reading/writing log files (ogit/Log)
     */
    //eslint-disable-next-line no-unused-vars
    writelog(logId, entries) {
        return Promise.reject(new Error('Log writing is not implemented yet'));
    }
    //eslint-disable-next-line no-unused-vars
    readlog(logId, options) {
        return Promise.reject(new Error('Log reading is not implemented yet'));
    }

    /**
     *  Returns previous versions of a vertex
     */
    history(
        id,
        {
            offset = false,
            limit = false,
            from = false,
            to = false,
            version = false,
            type = false,
            listMeta = false,
            includeDeleted = false,
            vid = false,
        } = {},
    ) {
        const headers = { 'ogit/_id': id, listMeta, includeDeleted };
        const body = {};

        if (offset !== false) {
            headers.offset = offset;
        }

        if (limit !== false) {
            headers.limit = limit;
        }

        if (from !== false) {
            headers.from = from;
        }

        if (to !== false) {
            headers.to = to;
        }

        if (version !== false) {
            headers.version = version;
        }

        if (type !== false) {
            headers.type = type;
        }

        if (vid !== false) {
            headers.vid = vid;
        }

        return this.dedupedRequest({
            type: 'history',
            headers: headers,
            body,
        });
    }

    /**
     *  Custom servlet endpoints
     *  always use HTTP transport, which knows how to handle these types.
     *
     *  The servletMethods are passed:
     *  - `fetch` a token/endpoint aware HTTP fetch API.
     *  - `options` the default fetch options you can override
     *  - `...args` the rest of the args passed in by the user for that method
     */
    addServlet(prefix, servletMethods, proxy) {
        if (!prefix) {
            throw new Error('[GRAPH] Must give prefix for servlet');
        }

        if (prefix in this) {
            throw new Error(
                '[GRAPH] Sevlet Extensions must have unique prefixes. Attempted to re-add `' +
                    prefix +
                    '`',
            );
        }

        const fetch = proxy ? this.proxyFetch(proxy) : this.fetch;
        const isFactory = typeof servletMethods === 'function';

        const servletDefinition = isFactory
            ? servletMethods(fetch, this.http.defaultOptions())
            : servletMethods;

        //create namespace.
        this[prefix] = {};
        //add servlet methods
        Object.keys(servletDefinition).reduce((acc, method) => {
            acc[method] = (...args) => {
                const servletMethod = servletDefinition[method];

                const legacyArgs = isFactory
                    ? []
                    : [fetch, this.http.defaultOptions()];

                const callArgs = [...legacyArgs, ...args];

                return servletMethod(...callArgs).catch((err) => {
                    //these are the special cases.
                    //regular errors end up with code === undefined, so not retryable.
                    switch (true) {
                        case isUnauthorized(err): //unauthorized (which means unauthenticated) invalidate TOKEN.
                            this.token.invalidate();
                            err.isRetryable = true;
                            break;
                        case isTransactionFail(err): //error persisting transaction. retryable.
                            err.isRetryable = true;
                            break;
                        //there are other known errors, e.g. 403, 400, etc... but they are not retryable.
                        default:
                            if (err.isRetryable === undefined) {
                                err.isRetryable = false;
                            }

                            break;
                    }

                    //a chance to retry - only once.
                    if (err.isRetryable) {
                        return servletMethod(...callArgs);
                    }

                    throw err;
                });
            };

            return acc;
        }, this[prefix]);

        //chainable? might be useful.
        return this;
    }

    getServlet(prefix) {
        return this[prefix];
    }
}
