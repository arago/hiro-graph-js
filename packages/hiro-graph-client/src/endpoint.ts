import qs, { ParsedUrlQueryInput } from 'querystring';

export const APIs = {
  http: {
    app: '/api/app/7.0',
    auth: '/api/auth/6.1',
    graph: '/api/graph/7.1',
    iam: '/api/iam/6.1',
    ki: '/api/ki/6',
    variables: '/api/variables/6',
  },
  ws: {
    events: '/api/events-ws/6.1',
    graph: '/api/graph-ws/6.1',
  },
};

export type WS_API = keyof typeof APIs['ws'];
export type HTTP_API = keyof typeof APIs['http'];

export class Endpoint<WS extends boolean = false> {
  private _value: string;
  protected readonly APIs?: Record<string, string>;

  constructor(value: string, isWebsocket?: WS) {
    this._value = value.replace(/\/$/, '');

    if (isWebsocket) {
      this._value = this._value.replace(/^http/, 'ws');
      this.APIs = APIs.ws;
    } else {
      this.APIs = APIs.http;
    }
  }

  get value() {
    return this._value;
  }

  api(
    name: WS extends true ? WS_API : HTTP_API,
    path?: string,
    query?: ParsedUrlQueryInput,
  ) {
    if (!this.APIs) {
      throw new Error(`No APIs defined for this endpoint: ${this.value}`);
    }

    let url = this._value + this.APIs[name];

    if (path) {
      if (path.indexOf('http') === 0) {
        url = path;
      } else {
        let p = path.replace(/\/$/, '');

        if (!p.startsWith('/')) {
          p = `/${p}`;
        }

        url += p;
      }
    }

    if (query && Object.keys(query).length > 0) {
      url += `/?${qs.stringify(query)}`;
    }

    return url;
  }
}
