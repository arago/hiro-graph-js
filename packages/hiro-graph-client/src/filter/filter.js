// @ts-nocheck
import { get, omit } from 'lodash';

/** Base Filter Constructor **/
export const Filter = function (attrib, comp, value) {
    this.type = 'filter';
    this.attrib = attrib;
    this.comp = comp;
    this.value = value;
};

Filter.prototype = {
    match: function (data) {
        var value = this.value;
        var attrv = get(data, this.attrib);

        switch (this.comp) {
            case '=':
                if (value === '*' && attrv) return true;
                else return Filter.matchString(attrv, value);
            case '<=':
                return Filter.matchLTE(attrv, value);
            case '>=':
                return Filter.matchGTE(attrv, value);
            case '~=':
                return Filter.matchApprox(attrv, value);
            case 'undefined':
                return !!attrv;
            default:
                throw new Error('Unknown comparison type');
        }
    },
    simplify: function () {
        if (this.filters) {
            if (this.filters.length === 1) return this.filters[0].simplify();
            else
                this.filters = this.filters.map(function (filter) {
                    return filter.simplify();
                });
        }

        return this;
    },
    _indent: function (indent, level, id_char) {
        var _i = parseInt(indent);

        if (indent === true) indent = Filter.indent;
        else if (!isNaN(_i)) indent = _i;
        else return '';

        if (id_char !== undefined && typeof id_char != 'string')
            throw new Error('Indent string must be string');

        level = level || 0;
        id_char = id_char || Filter.indent_char;

        return id_char.repeat(level * indent);
    },
    toString: function (indent, level, id_char) {
        return [
            this._indent(indent, level, id_char),
            '(',
            this.attrib,
            this.comp,
            this.value,
            ')',
        ].join('');
    },
};

/* A list of characters that need escaping */
Filter.escapeChars = ['*', '(', ')', '\\', String.fromCharCode(0)];

/* Default indent and character for beautify */
Filter.indent = 4;
Filter.indent_char = ' ';
Filter.collapse_not = true;

/* Transform event data to correct format for matching */
Filter.transformEvent = function (event) {
    return omit(
        {
            ...event,
            element: {
                ...(event.metadata || {}),
                ...(event.body || {}),
            },
            action: event.type,
        },
        ['body', 'type'],
    );
};

/* Escape a string value */
Filter.escape = function (value) {
    if (!value) return '';

    return value
        .split('')
        .map(function (c) {
            return Filter.escapeChars.indexOf(c) >= 0
                ? '\\' + c.charCodeAt(0).toString(16)
                : c;
        })
        .join('');
};

/* Unescape an escaped string value */
Filter.unescape = function (data) {
    var chars = data.split('');
    var out = [];
    var tmp;

    while (chars.length) {
        tmp = chars.shift();

        if (tmp === '\\') {
            tmp = chars.shift() + chars.shift();
            tmp = parseInt(tmp, 16);
            tmp = String.fromCharCode(tmp);
        }

        out.push(tmp);
    }

    return out.join('');
};

Filter.matchString = function (data, filter) {
    if (!data) return false;

    var match = Array.isArray(data) ? data : [data];

    if (filter.indexOf('*') < 0) {
        return match.some(function (cv) {
            if (cv) {
                return (
                    cv.toLowerCase() === Filter.unescape(filter).toLowerCase()
                );
            }

            return false;
        });
    }

    return Filter.matchSubstring(data, filter);
};

Filter.matchSubstring = function (data, filter) {
    var match = Array.isArray(data) ? data : [data];
    var pattern = filter.replace(/\*/g, '.*');
    pattern = pattern.replace(/\\([0-9a-fA-F]{2,2})/g, function (m, $1) {
        var s = String.fromCharCode(parseInt($1, 16));

        if (['(', ')', '\\', '*'].indexOf(s) >= 0) s = '\\x' + $1.toUpperCase();

        return s;
    });
    var regex = new RegExp('^' + pattern + '$', 'i');

    return match.some(function (cv) {
        return cv.match(regex);
    });
};

Filter.matchLTE = function (data, filter) {
    var match = Array.isArray(data) ? data : [data];

    return match.some(function (cv) {
        return cv <= filter;
    });
};

Filter.matchGTE = function (data, filter) {
    var match = Array.isArray(data) ? data : [data];

    return match.some(function (cv) {
        return cv >= filter;
    });
};

Filter.AND = function (filters) {
    return new GroupAnd(filters);
};

Filter.OR = function (filters) {
    return new GroupOr(filters);
};

Filter.NOT = function (filter) {
    if (!Array.isArray(filter)) filter = [filter];

    if (filter.length !== 1) throw new Error('NOT must wrap single filter');

    return new GroupNot(filter);
};

/** Grouping Constructor **/
export const Group = function (comp, filters) {
    this.type = 'group';
    this.comp = comp;
    this.filters = filters;
};

Group.prototype = Object.create(Filter.prototype);
Group.prototype.match = function (data) {
    return this._match(data);
};
Group.prototype.toString = function (indent, level, id_char) {
    level = level || 0;
    var id_str = this._indent(indent, level, id_char);
    var id_str2 = id_str;
    var nl = indent ? '\n' : '';

    if (Filter.collapse_not && this.comp === '!') {
        nl = '';
        id_str2 = '';
        indent = 0;
    }

    return [
        id_str,
        '(',
        this.comp,
        nl,
        this.filters
            .map(function (item) {
                return item.toString(indent, level + 1, id_char);
            })
            .join(nl),
        nl,
        id_str2,
        ')',
    ].join('');
};

var GroupOr = function (filters) {
    this.type = 'group';
    this.comp = '|';
    this.filters = filters;
};
GroupOr.prototype = Object.create(Group.prototype);
GroupOr.prototype._match = function (data) {
    return this.filters.some(function (cv, idx, all) {
        return cv.match(data);
    });
};

var GroupAnd = function (filters) {
    this.type = 'group';
    this.comp = '&';
    this.filters = filters;
};
GroupAnd.prototype = Object.create(Group.prototype);
GroupAnd.prototype._match = function (data) {
    return this.filters.every(function (cv, idx, all) {
        return cv.match(data);
    });
};

var GroupNot = function (filters) {
    this.type = 'group';
    this.comp = '!';
    this.filters = filters;
};
GroupNot.prototype = Object.create(Group.prototype);
GroupNot.prototype.simplify = function () {
    return this;
};
GroupNot.prototype._match = function (data) {
    return this.filters.every(function (cv, idx, all) {
        if (cv && typeof cv.match == 'function') {
            return !!!cv.match(data);
        } else {
            return false;
        }
    });
};

/** Attribute Constructor **/
export const Attribute = function (name) {
    this.name = name;
};

Attribute.prototype = {
    present: function () {
        return new Filter(this.name);
    },
    any: function () {
        return new Filter(this.name, '=', '*');
    },
    raw: function (value) {
        return new Filter(this.name, '=', value);
    },
    equalTo: function (value) {
        return new Filter(this.name, '=', this.escape(value));
    },
    endsWith: function (value) {
        return new Filter(this.name, '=', '*' + this.escape(value));
    },
    startsWith: function (value) {
        return new Filter(this.name, '=', this.escape(value) + '*');
    },
    contains: function (value) {
        return new Filter(this.name, '=', '*' + this.escape(value) + '*');
    },
    approx: function (value) {
        return new Filter(this.name, '~=', this.escape(value));
    },
    lte: function (value) {
        return new Filter(this.name, '<=', this.escape(value));
    },
    gte: function (value) {
        return new Filter(this.name, '>=', this.escape(value));
    },
    escapeChars: ['*', '(', ')', '\\', String.fromCharCode(0)],
    escape: function (value) {
        if (typeof value == 'number') value += '';

        var rv = [];

        for (var i = 0, l = value.length; i < l; i++) {
            rv.push(
                this.escapeChars.indexOf(value[i]) >= 0
                    ? '\\' + value.charCodeAt(i).toString(16)
                    : value[i],
            );
        }

        return rv.join('');
    },
};

Filter.attribute = function (name) {
    return new Attribute(name);
};
