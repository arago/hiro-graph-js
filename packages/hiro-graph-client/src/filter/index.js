export * from './filter';
export * from './parse';
