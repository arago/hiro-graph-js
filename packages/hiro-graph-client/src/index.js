import Token, { cannotGetToken } from './token';
import Client from './client';
import * as Errors from './errors';

export default Client;
export { Token, cannotGetToken, Errors };
