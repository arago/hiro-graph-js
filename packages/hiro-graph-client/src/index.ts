import { Token, cannotGetToken } from './token';
import { Client, ServletFactory } from './client';
import * as Errors from './errors';

export default Client;
export * from './gremlin';
export * from './lucene';
export * from './types';
export * from './jfilter';
export * from './endpoint';
export * as Servlets from './servlets/index';
export { cannotGetToken, Errors, ServletFactory, Token };
