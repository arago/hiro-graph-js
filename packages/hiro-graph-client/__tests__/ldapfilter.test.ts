import { GraphEventType } from '../src';
import { Filter, parseFilter } from '../src/filter';

test('LDAPFilter snapshots', () => {
  const tests = [
    Filter.AND([
      Filter.attribute('a').equalTo('b'),
      Filter.attribute('d').equalTo('e'),
    ]),
    Filter.OR([
      Filter.attribute('a').equalTo('b'),
      Filter.attribute('d').equalTo('e'),
    ]),
    Filter.AND([
      Filter.attribute('a').equalTo('b'),
      Filter.attribute('d').equalTo('e'),
    ]),
    Filter.AND([
      Filter.attribute('a').equalTo('b'),
      Filter.OR([Filter.attribute('e').equalTo('f')]),
      Filter.AND([
        Filter.attribute('d').equalTo('c'),
        Filter.attribute('g').equalTo('f'),
      ]),
    ]),
  ];

  tests.map((t) => expect(t.toString()).toMatchSnapshot());
});

test('LDAPFilter functions', () => {
  expect(Filter.attribute('a').equalTo('b').toString()).toBe('(a=b)');
  expect(
    Filter.AND([
      Filter.attribute('a').present(),
      Filter.attribute('b').present(),
    ]).toString(),
  ).toBe('(&(a)(b))');
  expect(
    Filter.OR([
      Filter.attribute('a').present(),
      Filter.attribute('b').present(),
    ]).toString(),
  ).toBe('(|(a)(b))');
});

test('LDAPFilter nested', () => {
  expect(
    Filter.AND([
      Filter.attribute('a').present(),
      Filter.AND([
        Filter.attribute('b').present(),
        Filter.attribute('c').present(),
      ]),
    ]).toString(),
  ).toBe('(&(a)(&(b)(c)))');
  expect(
    Filter.OR([
      Filter.attribute('a').present(),
      Filter.OR([
        Filter.attribute('b').present(),
        Filter.attribute('c').present(),
      ]),
    ]).toString(),
  ).toBe('(|(a)(|(b)(c)))');
});

test('LDAPFilter transform', () => {
  const filter = Filter.AND([
    Filter.attribute('element.ogit/_type').equalTo('ogit/Mobile/HealthInfo'),
    Filter.attribute('action').any(),
  ]);

  const res = {
    body: { 'ogit/_type': 'ogit/Mobile/HealthInfo' },
    type: 'UPDATE' as GraphEventType,
  };

  expect(filter.match(Filter.transformEvent(res))).toEqual(true);

  expect(
    Filter.transformEvent({ body: { 'ogit/_id': 'test' }, type: 'UPDATE' }),
  ).toEqual({
    element: { 'ogit/_id': 'test' },
    action: 'UPDATE',
  });
});

test('LDAPFilter test function', () => {
  expect(
    Filter.attribute('element.ogit/_id')
      .equalTo('test')
      .match({
        element: { 'ogit/_id': 'test' },
        action: 'UPDATE',
      }),
  ).toEqual(true);

  expect(
    Filter.attribute('action')
      .equalTo('UPDATE')
      .match({
        element: { 'ogit/_id': 'test' },
        action: 'CREATE',
      }),
  ).toEqual(false);

  expect(
    Filter.attribute('action')
      .any()
      .match({
        element: { 'ogit/_id': 'test' },
        action: 'CREATE',
      }),
  ).toEqual(true);

  expect(
    Filter.AND([
      Filter.attribute('element.ogit/_id').equalTo('test'),
      Filter.attribute('action').equalTo('CREATE'),
    ]).match({
      element: { 'ogit/_id': 'test' },
      action: 'CREATE',
    }),
  ).toEqual(true);

  expect(
    Filter.AND([
      Filter.attribute('element.ogit/_id').equalTo('test'),
      Filter.attribute('action').equalTo('CREATE'),
    ]).match({
      element: { 'ogit/_id': 'test2' },
      action: 'CREATE',
    }),
  ).toEqual(false);
});

test('LDAP Filter Parse Snapshot', () => {
  const tests = [
    `&(|(action = GET)(action = UPDATE)(action = CREATE)(action = DELETE))(vertex.ogit/_type = ogit/CustomApplicationData)(vertex./type = InstanceSettings)`,
    '(element.ogit/_type=ogit/Automation/AutomationIssue)',
  ];

  tests.map((t) => expect(parseFilter(t)).toMatchSnapshot());
});

test('LDAP Filter Parse', () => {
  expect(
    parseFilter('(element.ogit/_type=ogit/Automation/AutomationIssue)'),
  ).toEqual({
    type: 'filter',
    attrib: 'element.ogit/_type',
    comp: '=',
    value: 'ogit/Automation/AutomationIssue',
  });
});
