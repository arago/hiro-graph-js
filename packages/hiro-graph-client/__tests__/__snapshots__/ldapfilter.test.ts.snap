// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LDAP Filter Parse Snapshot 1`] = `
Object {
  "comp": "&",
  "filters": Array [
    Object {
      "comp": "|",
      "filters": Array [
        Object {
          "attrib": "action",
          "comp": "=",
          "type": "filter",
          "value": "GET",
        },
        Object {
          "attrib": "action",
          "comp": "=",
          "type": "filter",
          "value": "UPDATE",
        },
        Object {
          "attrib": "action",
          "comp": "=",
          "type": "filter",
          "value": "CREATE",
        },
        Object {
          "attrib": "action",
          "comp": "=",
          "type": "filter",
          "value": "DELETE",
        },
      ],
      "type": "group",
    },
    Object {
      "attrib": "vertex.ogit/_type",
      "comp": "=",
      "type": "filter",
      "value": "ogit/CustomApplicationData",
    },
    Object {
      "attrib": "vertex./type",
      "comp": "=",
      "type": "filter",
      "value": "InstanceSettings",
    },
  ],
  "type": "group",
}
`;

exports[`LDAP Filter Parse Snapshot 2`] = `
Object {
  "attrib": "element.ogit/_type",
  "comp": "=",
  "type": "filter",
  "value": "ogit/Automation/AutomationIssue",
}
`;

exports[`LDAPFilter snapshots 1`] = `"(&(a=b)(d=e))"`;

exports[`LDAPFilter snapshots 2`] = `"(|(a=b)(d=e))"`;

exports[`LDAPFilter snapshots 3`] = `"(&(a=b)(d=e))"`;

exports[`LDAPFilter snapshots 4`] = `"(&(a=b)(|(e=f))(&(d=c)(g=f)))"`;
